File: leetcode2904/leetcode2904.cpp
#include <bits/stdc++.h>
#include<string>
using namespace std;

class Solution {
    public:
        string shortestBeautifulSubstring(string s, int k) {
            if (count(s.begin(), s.end(), '1') < k) return "";
            string ans = s;
            int cnt = 0, left = 0;
            for (int right = 0; right < s.size(); right++){
                cnt += s[right] - '0';
                while (cnt > k || s[left] == '0')
                {
                    cnt -= s[left] - '0';
                    left ++;
                }
                if (cnt == k){
                    string t = s.substr(left, right - left + 1);
                    if (t.size() < ans.size() || t.size() == ans.size() && t < ans){
                        ans = t;
                    }
                }
            }
            return ans;
        }
        
    };

File: leetcode2904/leetcode2904.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        if s.count('1') < k:
            return ''
        ans = s
        cnt1 = left = 0
        for right, c in enumerate(s):
            cnt1 += int(c)
            while cnt1 > k or s[left] == '0':
                cnt1 -= int(s[left])
                left += 1
            if cnt1 == k:
                t = s[left: right + 1]
                if len(t) < len(ans) or len(t) == len(ans) and t < ans:
                    ans = t
        return ans

